machida3/machida.pony

```diff
@@ -552,7 +552,6 @@ primitive Machida
     recover val
       let pipeline_tree = PyPipelineTree(application_setup_data, env)
       let pipeline = pipeline_tree.build()?
-      Machida.dec_ref(application_setup_data)
```

Comment:
Why do we no longer need this `dec_ref`?

Reply:
This is the same solution implemented in https://github.com/WallarooLabs/wallaroo/pull/2684/files#diff-80d0b5a2ac917689f1a6180551b6c6aa for the `machida` case.

1. the `application_setup_data` object is created in `machida/main.pony` (and `machida3/main.pony`) when it executes the application module's `application_setup` function.
Since this object is created from pony, via the C API, that's the only reference to it.
2. Because we create the python step instances (computations, partitions, encoders, decoders, etc.) in python, as part of the `application_setup` function, their top-level reference is that list returned by `application_setup`.  We don't increment any references to them in pony. This means that once this object is collected, so will its child refs, including these step instances.
3. So when we decref this object here, and its ref count drops to 0, we enter a race with Python's GC on whether we finish serializing these objects before it collects them.

When the Python GC wins this race, we get the segfault we saw last week.

Reply:
Got it. Thanks for explanation.